modules/internal/String.chpl

```diff
@@ -192,6 +238,50 @@ module String {
       this.len = s_len;
     }
 
+    /*
+      :returns: The number of characters in the string.
+      */
+    inline proc length return len;
+
+    /*
+       Gets a version of the :record:`string` that is on the currently
+       executing locale.
+
+       :returns: A shallow copy if the :record:`string` is already on the
+                 current locale, otherwise a deep copy is performed.
+    */
+    inline proc localize() : string {
+      if _local || this.locale_id == chpl_nodeID {
+        return new string(this, owned=false);
+      } else {
+        const x:string = this; // assignment makes it local
+        return x;
+      }
+    }
+
+    /*
+      Get a :type:`c_string` from a :record:`string`.
+
+      .. warning::
+
+          This can only be called safely on a :record:`string` who's home is
+          the current locale.  This property can be enforced by calling
+          :proc:`string.localize()` before :proc:`~string.c_str()`.
+
+      For example:
+
+      .. code-block:: chapel
+
+          var my_string = "Hello!";
+          on different_locale {
+            printf("%s", my_string.localize().c_str());
+          }
+
+      :returns:
+          A :type:`c_string` that points to the underlying buffer used by this
+          :record:`string`.  The returned :type:`c_string` is only valid while
+          the :record:`string` is.
```

Comment:
The c_string returned is also only valid on the locale where it is created, correct?  More specifically, if you perform a localize call on the string and then generate a c_string, but then move to a different locale and try to access that c_string, you'd get garbage, right?  It might be worth mentioning that here as a reminder.